_posts/2020-09-03-final-report-hugo.md

```diff
@@ -0,0 +1,213 @@
+// to this (wow disgusting)
+md5_append(md5_state, reinterpret_cast<md5_byte_t*>(const_cast<char*>(str)), len);
+```
+
+There were loads and loads of these throughout the library, which hopefully look a lot more like C++ now.
+
+From here I started on modernisation of the library classes and constructs
+to modern C++. These progressed in [!28](https://gitlab.com/gromacs/tng/-/merge_requests/28) which outlined the structure we were aiming
+for, [!32](https://gitlab.com/gromacs/tng/-/merge_requests/32) which defined a specific API for blocks, [!35](https://gitlab.com/gromacs/tng/-/merge_requests/35) that updated the
+trajectory API and [!36]((https://gitlab.com/gromacs/tng/-/merge_requests/36))
+which progressed on the IO elements.
+
+An example of a modernisation change to the TNG API involves the use of templates to simplify the definition of TNG data blocks.
+A complicated `struct` that carried around a `void*` array for the data and a `tng_datatypes` enum to indicate the data type contained in the block could instead be a templated `class` as is shown below:
+
+```c++
+template<typename T> // type for data array
+TngDataBlock<T>::TngDataBlock(const int64_t&     id,
+                            // other stuff
+                              T*                 values) :
+```
+Template specialization is then used to do the correct operations for each datatype, reducing a lot of very repetitive code paths into a few simpler statements.
+
+
+Refactoring the TNG library was a huge task, much more so than intially anticipated. Due to this and some
+changing circumstances, the aforementioned modernisation MRs are all still open with more work required. I
+plan on working on these into the future but progress will be slow as I do not
+have solid blocks of time to dedicate. 
+
+All the work I have done on TNG itself can be found in my TNG
+[fork](https://gitlab.com/hugomacdermott/tng) as well as open PRs on the
+library [itself](https://gitlab.com/gromacs/tng).
+
+
+# PyTNG time!
+
+From here I switched focus to [PyTNG](https://github.com/MDAnalysis/pytng), a
+set of python bindings designed for use by MDAnalysis although technically a
+separate library. I worked on PyTNG for most of the second half of GSOC. This required
+changing gears a little bit as well as learning Cython, which was initially a
+bit of a learning curve for me. 
+
+Firstly I changed the TNG libraries exported with PyTNG itself to
+include the bugfixes obtained as part of the earlier improvements to the TNG
+library. This was incorporated as part of
+[#28](https://github.com/MDAnalysis/pytng/pull/28). I then improved the TNG library
+calls in PyTNG in [#29](https://github.com/MDAnalysis/pytng/pull/29).
+
+Following some design discussions, we then moved towards a newer design for the
+bindings, so as to be able to read all the blocks available in a TNG file with
+maximal speed. This was acheived in merging
+[#32](https://github.com/MDAnalysis/pytng/pull/32) and ongoing work in
```

Comment:
If you want to re-use links then you can do this markdown with
```
[#32][PR32] and [#38][PR38], which are total


[#32]: https://github.com/MDAnalysis/pytng/pull/32
[#38]: https://github.com/MDAnalysis/pytng/pull/38
```
See https://github.github.com/gfm/#reference-link